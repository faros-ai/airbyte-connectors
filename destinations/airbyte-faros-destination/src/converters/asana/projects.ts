import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';

import {DestinationModel, DestinationRecord} from '../converter';
import {AsanaCommon, AsanaConverter} from './common';

export class Projects extends AsanaConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_Project'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const project = record.record.data;

    const tmsProject: DestinationRecord = {
      model: 'tms_Project',
      record: {
        uid: project.gid,
        name: project.name,
        description: project.notes?.substring(
          0,
          AsanaCommon.MAX_DESCRIPTION_LENGTH
        ),
        createdAt: Utils.toDate(project.created_at),
        updatedAt: Utils.toDate(project.modified_at),
        source,
      },
    };

    return [tmsProject];
  }
}
