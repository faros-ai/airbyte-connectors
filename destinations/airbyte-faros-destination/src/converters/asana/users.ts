import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AsanaCommon, AsanaConverter, AsanaUser} from './common';

export class Users extends AsanaConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_User'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const user = record.record.data as AsanaUser;

    return [AsanaCommon.tms_User(user, source)];
  }
}
