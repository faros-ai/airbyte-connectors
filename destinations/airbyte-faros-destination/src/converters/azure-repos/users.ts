import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureReposConverter} from './common';
import {User, UserType, UserTypeCategory} from './models';

export class Users extends AzureReposConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'vcs_Membership',
    'vcs_User',
  ];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const userItem = record.record.data as User;
    const res: DestinationRecord[] = [];
    const organizationName = this.getOrganizationFromUrl(
      userItem._links.self.href
    );
    const organization = {uid: organizationName, source};
    const type: UserType = {
      category: UserTypeCategory.User,
      detail: userItem.subjectKind,
    };
    res.push({
      model: 'vcs_Membership',
      record: {
        organization,
        user: {uid: userItem.principalName, source},
      },
    });

    res.push({
      model: 'vcs_User',
      record: {
        uid: userItem.principalName,
        name: userItem.displayName,
        type,
        htmlUrl: userItem.url,
        source,
      },
    });
    return res;
  }
}
