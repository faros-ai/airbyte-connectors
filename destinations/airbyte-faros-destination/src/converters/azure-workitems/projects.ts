import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureWorkitemsConverter} from './common';
import {Project} from './models';

export class Projects extends AzureWorkitemsConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_Project'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const project = record.record.data as Project;
    return [
      {
        model: 'tms_Project',
        record: {
          uid: String(project.id),
          name: project.name,
          description: Utils.cleanAndTruncate(project.description),
          updatedAt: Utils.toDate(project.lastUpdateTime),
          source: this.streamName.source,
        },
      },
    ];
  }
}
