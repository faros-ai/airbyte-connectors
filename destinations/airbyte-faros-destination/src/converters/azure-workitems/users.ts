import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureWorkitemsConverter} from './common';
import {User} from './models';

export class Users extends AzureWorkitemsConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_User'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const userItem = record.record.data as User;
    const res: DestinationRecord[] = [];

    res.push({
      model: 'tms_User',
      record: {
        uid: userItem.principalName,
        name: userItem.displayName,
        emailAddress: userItem.mailAddress,
        source,
      },
    });
    return res;
  }
}
