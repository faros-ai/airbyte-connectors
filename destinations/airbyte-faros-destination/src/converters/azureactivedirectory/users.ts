import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureActiveDirectoryConverter} from './common';
import {User} from './models';

export class Users extends AzureActiveDirectoryConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'geo_Address',
    'geo_Location',
    'identity_Identity',
    'org_Department',
    'org_Employee',
  ];

  private seenDepartments = new Set<string>();

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const user = record.record.data as User;
    const joinedAt = Utils.toDate(user.createdDateTime);
    const manager = user.manager ? {uid: user.manager, source} : undefined;
    const uid = user.id;
    const res: DestinationRecord[] = [];

    if (user.department && !this.seenDepartments.has(user.department)) {
      this.seenDepartments.add(user.department);
      res.push({
        model: 'org_Department',
        record: {
          uid: user.department,
          name: user.department,
        },
      });
    }

    res.push({
      model: 'identity_Identity',
      record: {
        uid,
        fullName: `${user.givenName} ${user.surname}`,
        lastName: user.surname,
        primaryEmail: user.mail,
        emails: [user.mail],
      },
    });

    const location = user.streetAddress
      ? {uid: user.streetAddress, source}
      : undefined;

    if (user.streetAddress) {
      res.push({
        model: 'geo_Address',
        record: {
          uid: user.streetAddress,
          fullAddress: user.streetAddress,
          street: user.streetAddress,
          postalCode: user.postalCode,
        },
      });
      const geo_Location = {
        uid: user.streetAddress,
        name: user.streetAddress,
        raw: user.streetAddress,
        address: {uid: user.streetAddress},
      };
      res.push({model: 'geo_Location', record: geo_Location});
    }
    res.push({
      model: 'org_Employee',
      record: {
        uid,
        title: user.displayName,
        level: 0,
        joinedAt,
        department: user.department ? {uid: user.department} : null,
        identity: {uid, source},
        manager,
        location,
        source,
      },
    });
    return res;
  }
}
