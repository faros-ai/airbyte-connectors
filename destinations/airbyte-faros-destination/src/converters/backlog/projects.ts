import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';

import {DestinationModel, DestinationRecord, StreamContext} from '../converter';
import {BacklogCommon, BacklogConverter} from './common';
import {Project} from './models';

export class Projects extends BacklogConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'tms_Project',
    'tms_Release',
    'tms_Sprint',
    'tms_TaskBoard',
    'tms_TaskBoardProjectRelationship',
  ];
  async convert(
    record: AirbyteRecord,
    ctx: StreamContext
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const project = record.record.data as Project;
    const uid = String(project.id);
    const res: DestinationRecord[] = [];

    const maxDescriptionLength = this.maxDescriptionLength(ctx);

    res.push({
      model: 'tms_Project',
      record: {
        uid,
        name: project.name,
        description: null,
        source,
      },
    });
    res.push({
      model: 'tms_TaskBoard',
      record: {
        uid,
        name: project.name,
        source,
      },
    });
    res.push({
      model: 'tms_TaskBoardProjectRelationship',
      record: {
        board: {uid, source},
        project: {uid, source},
      },
    });
    for (const versionMilestone of project.versionMilestones) {
      res.push({
        model: 'tms_Sprint',
        record: {
          uid: String(versionMilestone.id),
          name: versionMilestone.name,
          description: versionMilestone.description?.substring(
            0,
            maxDescriptionLength
          ),
          startedAt: Utils.toDate(versionMilestone.startDate),
          endedAt: Utils.toDate(versionMilestone.releaseDueDate),
          state: BacklogCommon.getSprintState(versionMilestone),
          source,
        },
      });
      if (versionMilestone.releaseDueDate) {
        res.push({
          model: 'tms_Release',
          record: {
            uid: String(versionMilestone.id),
            name: versionMilestone.name,
            description: versionMilestone.description?.substring(
              0,
              maxDescriptionLength
            ),
            startedAt: Utils.toDate(versionMilestone.startDate),
            releasedAt: Utils.toDate(versionMilestone.releaseDueDate),
            source,
          },
        });
      }
    }
    return res;
  }
}
