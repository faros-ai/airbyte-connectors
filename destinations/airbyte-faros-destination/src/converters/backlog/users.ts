import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {BacklogConverter} from './common';
import {User} from './models';

export class Users extends BacklogConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_User'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const user = record.record.data as User;
    return [
      {
        model: 'tms_User',
        record: {
          uid: String(user.id),
          name: user.name,
          emailAddress: user.mailAddress,
          source,
        },
      },
    ];
  }
}
