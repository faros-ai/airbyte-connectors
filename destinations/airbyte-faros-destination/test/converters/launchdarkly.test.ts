import {AirbyteRecord} from 'faros-airbyte-cdk';
import {
  generateBasicTestSuite,
  initMockttp,
  tempConfig,
} from 'faros-airbyte-testing-tools';
import {getLocal} from 'mockttp';

import {Environments} from '../../src/converters/launchdarkly/environments';
import {Experiments} from '../../src/converters/launchdarkly/experiments';
import {FeatureFlags} from '../../src/converters/launchdarkly/feature_flags';
import {Projects} from '../../src/converters/launchdarkly/projects';
import {Users} from '../../src/converters/launchdarkly/users';

describe('launchdarkly', () => {
  const mockttp = getLocal({debug: false, recordTraffic: false});
  let configPath: string;

  beforeEach(async () => {
    await initMockttp(mockttp);
    configPath = await tempConfig({api_url: mockttp.url});
  });

  afterEach(async () => {
    await mockttp.stop();
  });

  generateBasicTestSuite({sourceName: 'launchdarkly'});

  describe('projects', () => {
    const converter = new Projects();
    const PROJECT = {
      key: 'test-project',
      name: 'Test Project',
      tags: ['test', 'automated'],
    };

    test('basic project', async () => {
      const record = AirbyteRecord.make('projects', PROJECT);
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });

    test('project without tags', async () => {
      const record = AirbyteRecord.make('projects', {
        key: 'simple-project',
        name: 'Simple Project',
      });
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });
  });

  describe('environments', () => {
    const converter = new Environments();
    const ENVIRONMENT = {
      key: 'production',
      name: 'Production',
      color: 'FF0000',
      tags: ['prod'],
    };

    test('basic environment', async () => {
      const record = AirbyteRecord.make('environments', ENVIRONMENT);
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });

    test('environment without color and tags', async () => {
      const record = AirbyteRecord.make('environments', {
        key: 'staging',
        name: 'Staging',
      });
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });
  });

  describe('feature_flags', () => {
    const converter = new FeatureFlags();
    const FEATURE_FLAG = {
      key: 'test-flag',
      name: 'Test Flag',
      kind: 'boolean',
      description: 'A test feature flag',
      tags: ['test'],
    };

    test('basic feature flag', async () => {
      const record = AirbyteRecord.make('feature_flags', FEATURE_FLAG);
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });

    test('multivariate feature flag', async () => {
      const record = AirbyteRecord.make('feature_flags', {
        key: 'multivariate-flag',
        name: 'Multivariate Flag',
        kind: 'multivariate',
        description: 'A multivariate feature flag',
        tags: ['test', 'multivariate'],
      });
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });

    test('feature flag without description and tags', async () => {
      const record = AirbyteRecord.make('feature_flags', {
        key: 'simple-flag',
        name: 'Simple Flag',
        kind: 'boolean',
      });
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });
  });

  describe('users', () => {
    const converter = new Users();
    const USER = {
      key: 'user-123',
      name: 'John Doe',
      email: 'john@example.com',
      country: 'US',
    };

    test('basic user', async () => {
      const record = AirbyteRecord.make('users', USER);
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });

    test('user with minimal data', async () => {
      const record = AirbyteRecord.make('users', {
        key: 'user-456',
      });
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });

    test('user without email and country', async () => {
      const record = AirbyteRecord.make('users', {
        key: 'user-789',
        name: 'Jane Smith',
      });
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });
  });

  describe('experiments', () => {
    const converter = new Experiments();
    const EXPERIMENT = {
      key: 'test-experiment',
      name: 'Test Experiment',
      description: 'A test experiment',
      hypothesis: 'This will improve engagement',
    };

    test('basic experiment', async () => {
      const record = AirbyteRecord.make('experiments', EXPERIMENT);
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });

    test('experiment without description and hypothesis', async () => {
      const record = AirbyteRecord.make('experiments', {
        key: 'simple-experiment',
        name: 'Simple Experiment',
      });
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });
  });
});
