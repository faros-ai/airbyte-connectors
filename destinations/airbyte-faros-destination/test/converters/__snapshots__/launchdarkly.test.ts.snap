// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`launchdarkly environments basic environment 1`] = `
Array [
  Object {
    "model": "ffs_Environment",
    "record": Object {
      "color": "FF0000",
      "name": "Production",
      "source": "LaunchDarkly",
      "tags": Array [
        "prod",
      ],
      "uid": "production",
    },
  },
]
`;

exports[`launchdarkly environments environment without color and tags 1`] = `
Array [
  Object {
    "model": "ffs_Environment",
    "record": Object {
      "color": undefined,
      "name": "Staging",
      "source": "LaunchDarkly",
      "tags": Array [],
      "uid": "staging",
    },
  },
]
`;

exports[`launchdarkly experiments basic experiment 1`] = `
Array [
  Object {
    "model": "ffs_Experiment",
    "record": Object {
      "description": "A test experiment",
      "hypothesis": "This will improve engagement",
      "name": "Test Experiment",
      "source": "LaunchDarkly",
      "uid": "test-experiment",
    },
  },
]
`;

exports[`launchdarkly experiments experiment without description and hypothesis 1`] = `
Array [
  Object {
    "model": "ffs_Experiment",
    "record": Object {
      "description": undefined,
      "hypothesis": undefined,
      "name": "Simple Experiment",
      "source": "LaunchDarkly",
      "uid": "simple-experiment",
    },
  },
]
`;

exports[`launchdarkly feature_flags basic feature flag 1`] = `
Array [
  Object {
    "model": "ffs_FeatureFlag",
    "record": Object {
      "description": "A test feature flag",
      "kind": "boolean",
      "name": "Test Flag",
      "source": "LaunchDarkly",
      "tags": Array [
        "test",
      ],
      "uid": "test-flag",
    },
  },
]
`;

exports[`launchdarkly feature_flags feature flag without description and tags 1`] = `
Array [
  Object {
    "model": "ffs_FeatureFlag",
    "record": Object {
      "description": undefined,
      "kind": "boolean",
      "name": "Simple Flag",
      "source": "LaunchDarkly",
      "tags": Array [],
      "uid": "simple-flag",
    },
  },
]
`;

exports[`launchdarkly feature_flags multivariate feature flag 1`] = `
Array [
  Object {
    "model": "ffs_FeatureFlag",
    "record": Object {
      "description": "A multivariate feature flag",
      "kind": "multivariate",
      "name": "Multivariate Flag",
      "source": "LaunchDarkly",
      "tags": Array [
        "test",
        "multivariate",
      ],
      "uid": "multivariate-flag",
    },
  },
]
`;

exports[`launchdarkly launchdarkly basic test process records from all streams 1`] = `Array []`;

exports[`launchdarkly projects basic project 1`] = `
Array [
  Object {
    "model": "ffs_Project",
    "record": Object {
      "name": "Test Project",
      "source": "LaunchDarkly",
      "tags": Array [
        "test",
        "automated",
      ],
      "uid": "test-project",
    },
  },
]
`;

exports[`launchdarkly projects project without tags 1`] = `
Array [
  Object {
    "model": "ffs_Project",
    "record": Object {
      "name": "Simple Project",
      "source": "LaunchDarkly",
      "tags": Array [],
      "uid": "simple-project",
    },
  },
]
`;

exports[`launchdarkly users basic user 1`] = `
Array [
  Object {
    "model": "ffs_User",
    "record": Object {
      "country": "US",
      "email": "john@example.com",
      "name": "John Doe",
      "source": "LaunchDarkly",
      "uid": "user-123",
    },
  },
]
`;

exports[`launchdarkly users user with minimal data 1`] = `
Array [
  Object {
    "model": "ffs_User",
    "record": Object {
      "country": undefined,
      "email": undefined,
      "name": undefined,
      "source": "LaunchDarkly",
      "uid": "user-456",
    },
  },
]
`;

exports[`launchdarkly users user without email and country 1`] = `
Array [
  Object {
    "model": "ffs_User",
    "record": Object {
      "country": undefined,
      "email": undefined,
      "name": "Jane Smith",
      "source": "LaunchDarkly",
      "uid": "user-789",
    },
  },
]
`;
